app/javascript/lib/validation/validate-file.js

```diff
@@ -0,0 +1,166 @@
+/**
+ * @fileoverview Client-side file validation (CSFV) for upload and sync UI
+ */
+
+import { GiB, MiB } from 'lib/validation/io'
+import { parseFile } from './validate-file-content'
+import { logFileValidation } from './log-validation'
+import { fetchBucketFile } from 'lib/scp-api'
+
+
+/** take an array of [category, type, msg] issues, and format it */
+function formatIssues(issues) {
+  const errors = issues.filter(issue => issue[0] === 'error')
+  const warnings = issues.filter(issue => issue[0] === 'warn')
+
+  return { errors, warnings }
+}
+
+/** Validate name uniqueness and file extension */
+function validateFileName(file, studyFile, allStudyFiles, allowedFileExts=['*']) {
+  const otherFiles = allStudyFiles.filter(f => f._id != studyFile._id)
+  const otherNames = otherFiles.map(f => f.name)
+  const otherUploadFileNames = otherFiles.map(f => f.upload_file_name)
+  const otherLocalFileNames = otherFiles.map(f => f.uploadSelection?.name)
+
+  const issues = []
+
+  if (otherNames.includes(file.name) ||
+    otherUploadFileNames.includes(file.name) ||
+    otherLocalFileNames.includes(file.name)) {
+    const msg = `A file named ${file.name} already exists in your study`
+    issues.push(['error', 'filename:duplicate', msg])
+  }
+
+  if (!allowedFileExts.includes('*') && !allowedFileExts.some(ext => file.name.endsWith(ext))) {
+    const msg = `Allowed extensions are ${allowedFileExts.join(' ')}`
+    issues.push(['error', 'filename:extension', msg])
+  }
+
+  return issues
+}
+
+/** checks name uniqueness, file extension, and then file content.
+ * The first two checks short-circuit the latter, since those will often just mean the user
+ * has local the wrong file, and showing a string of validation errors would be confusing
+ * @param file {File} the File object from the input
+ * @param studyFile {StudyFile} the JS object corresponding to the StudyFile
+ * @param allStudyFiles {StudyFile[]} the array of all files for the study, used for name uniqueness checks
+ * @param allowedFileExts { String[] } array of allowable extensions, ['*'] for all
+ */
+export async function validateLocalFile(file, studyFile, allStudyFiles=[], allowedFileExts=['*']) {
+  const nameIssues = validateFileName(file, studyFile, allStudyFiles, allowedFileExts)
+
+  let issuesObj
+  if (nameIssues.length === 0) {
+    const fileOptions = {
+      use_metadata_convention: studyFile.use_metadata_convention
+    }
+    const studyFileType = studyFile.file_type
+    const { fileInfo, issues, perfTime } = await parseFile(file, studyFileType, fileOptions)
+
+    const allIssues = issues.concat(nameIssues)
+    issuesObj = formatIssues(allIssues)
+
+    const perfTimes = {
+      perfTime,
+      'perfTime:parseFile': perfTime
+    }
+
+    logFileValidation(fileInfo, issuesObj, perfTimes)
+  } else {
+    issuesObj = formatIssues(nameIssues)
+  }
+
+  if (file.size >= GiB) {
+    issuesObj.infos = [['info', 'size:large', '']]
+  }
```

Comment:
being able to add this here so cleanly indicates to me that you did a nice job refactoring the local/remote validation logic

Reply:
Thanks!  The refactor largely centered around this specific line.